File: lib/features/dictionary/domain/dictionary_search_service.dart
import 'package:taigi_dict/features/dictionary/dictionary.dart';

List<Map<String, Object>> buildDictionarySearchIndex(
  List<DictionaryEntry> entries,
) {
  return entries
      .map((entry) {
        return <String, Object>{
          'id': entry.id,
          'headwords': _headwordFieldsForEntry(entry),
          'definitions': _definitionFieldsForEntry(entry),
        };
      })
      .toList(growable: false);
}

List<int> searchDictionaryEntryIds(
  List<Map<String, Object>> searchIndex,
  String query,
) {
  final matched = <_ScoredSearchHit>[];
  for (final searchRow in searchIndex) {
    final match = _matchSearchRow(searchRow, query);
    if (match != null) {
      matched.add(match);
    }
  }

  matched.sort((left, right) {
    final comparePriority = left.score.compareTo(right.score);
    if (comparePriority != 0) {
      return comparePriority;
    }

    final compareLength = left.matchedLength.compareTo(right.matchedLength);
    if (compareLength != 0) {
      return compareLength;
    }

    return left.entryId.compareTo(right.entryId);
  });

  return matched.take(60).map((item) => item.entryId).toList(growable: false);
}

_ScoredSearchHit? _matchSearchRow(Map<String, Object> searchRow, String query) {
  final entryId = searchRow['id'] as int;
  final headwordMatch = _bestMatchLength(
    (searchRow['headwords'] as List<Object>).cast<String>(),
    query,
  );
  if (headwordMatch != null) {
    final score = headwordMatch == query.length ? 0 : 1;
    return _ScoredSearchHit(entryId, score, headwordMatch);
  }

  final definitionMatch = _bestMatchLength(
    (searchRow['definitions'] as List<Object>).cast<String>(),
    query,
  );
  if (definitionMatch == null) {
    return null;
  }
  return _ScoredSearchHit(entryId, 2, definitionMatch);
}

List<String> _headwordFieldsForEntry(DictionaryEntry entry) {
  final fields = <String>{};
  final hanji = normalizeQuery(entry.hanji);
  if (hanji.isNotEmpty) {
    fields.add(hanji);
  }
  final romanization = normalizeQuery(entry.romanization);
  if (romanization.isNotEmpty) {
    fields.add(romanization);
  }
  return fields.toList(growable: false);
}

List<String> _definitionFieldsForEntry(DictionaryEntry entry) {
  return entry.senses
      .map((sense) => normalizeQuery(sense.definition))
      .where((definition) => definition.isNotEmpty)
      .toSet()
      .toList(growable: false);
}

int? _bestMatchLength(List<String> fields, String query) {
  int? bestLength;
  for (final field in fields) {
    if (field.isEmpty || query.isEmpty || !field.contains(query)) {
      continue;
    }
    if (bestLength == null || field.length < bestLength) {
      bestLength = field.length;
    }
  }
  return bestLength;
}

class _ScoredSearchHit {
  const _ScoredSearchHit(this.entryId, this.score, this.matchedLength);

  final int entryId;
  final int score;
  final int matchedLength;
}

String normalizeQuery(String input) {
  var normalized = removeTones(input.trim());
  normalized = normalized.replaceAll(RegExp(r'[1-8]'), '');
  normalized = normalized.replaceAll(RegExp(r'\s+'), ' ').trim();
  normalized = normalized.replaceAll(RegExp(r'[-_/]'), ' ');
  normalized = normalized.replaceAll(RegExp("[【】\\[\\]（）()、,.;:!?\"'`]+"), ' ');
  normalized = normalized.replaceAll(RegExp(r'\s+'), ' ').trim();
  return normalized;
}

String removeTones(String input) {
  var normalized = input.toLowerCase();
  for (final entry in _romanizationFold.entries) {
    normalized = normalized.replaceAll(entry.key, entry.value);
  }
  normalized = normalized.replaceAll(RegExp(r'[\u0300-\u036f]'), '');
  normalized = normalized.replaceAll('o͘', 'oo');
  normalized = normalized.replaceAll('ⁿ', 'n');
  normalized = normalized.replaceAll(RegExp(r'[1-8]'), '');
  return normalized;
}

const Map<String, String> _romanizationFold = {
  'á': 'a',
  'à': 'a',
  'â': 'a',
  'ǎ': 'a',
  'ā': 'a',
  'ä': 'a',
  'ã': 'a',
  'é': 'e',
  'è': 'e',
  'ê': 'e',
  'ē': 'e',
  'ë': 'e',
  'í': 'i',
  'ì': 'i',
  'î': 'i',
  'ī': 'i',
  'ï': 'i',
  'ó': 'o',
  'ò': 'o',
  'ô': 'o',
  'ō': 'o',
  'ö': 'o',
  'ő': 'o',
  'ú': 'u',
  'ù': 'u',
  'û': 'u',
  'ū': 'u',
  'ü': 'u',
  'ḿ': 'm',
  'm̀': 'm',
  'm̂': 'm',
  'ń': 'n',
  'ǹ': 'n',
  'n̂': 'n',
};

File: swift-native/Sources/TaigiDictCore/Domain/Dictionary/DictionarySearchService.swift
import Foundation

public struct DictionarySearchRow: Sendable {
    public var entryID: Int64
    public var headwords: [String]
    public var definitions: [String]

    public init(entryID: Int64, headwords: [String], definitions: [String]) {
        self.entryID = entryID
        self.headwords = headwords
        self.definitions = definitions
    }
}

public enum DictionarySearchService {
    public static let defaultLimit = 60

    public static func buildSearchIndex(entries: [DictionaryEntry]) -> [DictionarySearchRow] {
        entries.map { entry in
            DictionarySearchRow(
                entryID: entry.id,
                headwords: headwordFields(for: entry),
                definitions: definitionFields(for: entry)
            )
        }
    }

    public static func searchEntryIDs(
        index: [DictionarySearchRow],
        query rawQuery: String,
        limit: Int = defaultLimit
    ) -> [Int64] {
        let query = TextNormalization.normalizeQuery(rawQuery)
        guard !query.isEmpty else {
            return []
        }

        return index
            .compactMap { match(row: $0, query: query) }
            .sorted()
            .prefix(limit)
            .map(\.entryID)
    }

    private static func match(row: DictionarySearchRow, query: String) -> ScoredSearchHit? {
        if let headwordMatch = bestMatchLength(fields: row.headwords, query: query) {
            let score = headwordMatch == query.count ? 0 : 1
            return ScoredSearchHit(entryID: row.entryID, score: score, matchedLength: headwordMatch)
        }

        guard let definitionMatch = bestMatchLength(fields: row.definitions, query: query) else {
            return nil
        }

        return ScoredSearchHit(entryID: row.entryID, score: 2, matchedLength: definitionMatch)
    }

    private static func headwordFields(for entry: DictionaryEntry) -> [String] {
        uniqueNonEmpty([
            TextNormalization.normalizeQuery(entry.hanji),
            TextNormalization.normalizeQuery(entry.romanization),
        ])
    }

    private static func definitionFields(for entry: DictionaryEntry) -> [String] {
        uniqueNonEmpty(entry.senses.map { TextNormalization.normalizeQuery($0.definition) })
    }

    private static func bestMatchLength(fields: [String], query: String) -> Int? {
        fields.reduce(nil) { bestLength, field in
            guard !field.isEmpty, !query.isEmpty, field.contains(query) else {
                return bestLength
            }

            if let bestLength {
                return min(bestLength, field.count)
            }

            return field.count
        }
    }

    private static func uniqueNonEmpty(_ values: [String]) -> [String] {
        var seen = Set<String>()
        var result: [String] = []

        for value in values where !value.isEmpty && seen.insert(value).inserted {
            result.append(value)
        }

        return result
    }
}

private struct ScoredSearchHit: Comparable {
    var entryID: Int64
    var score: Int
    var matchedLength: Int

    static func < (left: ScoredSearchHit, right: ScoredSearchHit) -> Bool {
        if left.score != right.score {
            return left.score < right.score
        }

        if left.matchedLength != right.matchedLength {
            return left.matchedLength < right.matchedLength
        }

        return left.entryID < right.entryID
    }
}
